### ISSUE TITLE
git fetch --deepen stuck on deepeing

### ISSUE THREAD

**Issue by nhatkhai** [git-for-windows/git#3192]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.30.2.windows.1
cpu: x86_64
built from commit: 37f0a4afad027cf030747a1ef45537ba86162960
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.17763.1821]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: Nano
Custom Editor Path:
Default Branch Option:
Path Option: CmdTools
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFCommitAsIs
Bash Terminal Option: ConHost
Git Pull Behavior Option: Merge
Use Credential Manager: Disabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Enabled
Enable Pseudo Console Support: Enabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

NONE

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Git CMD

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
(1)$ git lfs install

(2)$ git clone ssh://git@XYZ/ABC/REPO.git my-reference

(3)$ git clone --depth=1 ssh://git@XYZ/ABC/REPO.git --reference ./my-reference

(4)$ cd REPO

(5)$ git fetch --deepen 600
16:48:50.443646 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/bin
16:48:50.452433 git.c:444               trace: built-in: git fetch --deepen 600
16:48:50.458292 run-command.c:664       trace: run_command: unset GIT_PREFIX; GIT_PROTOCOL=version=2 ssh -o SendEnv=GIT_PROTOCOL -p 7999 git@XYZ 'git-upload-pack '\''/ABC/REPO.git'\'''
16:48:52.221755 run-command.c:664       trace: run_command: unset GIT_PREFIX; git --git-dir=/my-reference for-each-ref '--format=%(objectname)'
16:48:52.318424 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
16:48:52.334083 git.c:444               trace: built-in: git for-each-ref '--format=%(objectname)'
remote: Enumerating objects: 6068, done.
remote: Counting objects: 100% (5662/5662), done.
remote: Compressing objects: 100% (1701/1701), done.
16:48:52.797220 run-command.c:664       trace: run_command: git --shallow-file c:/projects/REPO/.git/shallow.lock index-pack --stdin -v --fix-thin '--keep=fetch-pack 203100 on MY-PC1' --pack_header=2,5238
16:48:52.874338 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
16:48:52.883147 git.c:444               trace: built-in: git index-pack --stdin -v --fix-thin '--keep=fetch-pack 203100 on MY-PC1' --pack_header=2,5238
remote: Total 5238 (delta 3500), reused 4990 (delta 3285)
Receiving objects: 100% (5238/5238), 9.29 MiB | 1.08 MiB/s, done.
Resolving deltas: 100% (3500/3500), completed with 243 local objects.
16:49:02.522004 run-command.c:664       trace: run_command: git --shallow-file c:/projects/REPO/.git/shallow.lock rev-list --objects --stdin --quiet --alternate-refs
16:49:02.601096 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
16:49:02.609884 git.c:444               trace: built-in: git rev-list --objects --stdin --quiet --alternate-refs
16:49:02.657117 run-command.c:664       trace: run_command: unset GIT_PREFIX; git --git-dir=/my-reference for-each-ref '--format=%(objectname)'
16:49:02.757829 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
16:49:02.770090 git.c:444               trace: built-in: git for-each-ref '--format=%(objectname)'
16:49:03.581331 run-command.c:1625      run_processes_parallel: preparing to run up to 1 tasks
16:49:03.581331 run-command.c:1657      run_processes_parallel: done
16:49:03.582306 run-command.c:664       trace: run_command: git maintenance run --auto --no-quiet
16:49:03.666289 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
16:49:03.675068 git.c:444               trace: built-in: git maintenance run --auto --no-quiet

(6)$ git rev-list --count HEAD
624

(7)$ git fetch --deepen 600
16:54:09.793892 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/bin
16:54:09.802696 git.c:444               trace: built-in: git fetch --deepen 600
16:54:09.808551 run-command.c:664       trace: run_command: unset GIT_PREFIX; GIT_PROTOCOL=version=2 ssh -o SendEnv=GIT_PROTOCOL -p 7999 git@XYZ 'git-upload-pack '\''/ABC/REPO.git'\'''
16:54:11.554046 run-command.c:664       trace: run_command: unset GIT_PREFIX; git --git-dir=//att-repos/git_repositories/caches for-each-ref '--format=%(objectname)'
16:54:11.649753 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
16:54:11.666338 git.c:444               trace: built-in: git for-each-ref '--format=%(objectname)'
16:54:12.141794 run-command.c:664       trace: run_command: git --shallow-file c:/projects/REPO/.git/shallow.lock unpack-objects --pack_header=2,0
remote: Total 0 (delta 0), reused 0 (delta 0)
16:54:12.224791 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
16:54:12.232603 git.c:444               trace: built-in: git unpack-objects --pack_header=2,0
16:54:12.249208 run-command.c:664       trace: run_command: git --shallow-file c:/projects/REPO/.git/shallow.lock rev-list --objects --stdin --quiet --alternate-refs
16:54:12.328300 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
16:54:12.336113 git.c:444               trace: built-in: git rev-list --objects --stdin --quiet --alternate-refs
16:54:12.378098 run-command.c:664       trace: run_command: unset GIT_PREFIX; git --git-dir=//att-repos/git_repositories/caches for-each-ref '--format=%(objectname)'
16:54:12.458168 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
16:54:12.469884 git.c:444               trace: built-in: git for-each-ref '--format=%(objectname)'
16:54:13.355240 run-command.c:1625      run_processes_parallel: preparing to run up to 1 tasks
16:54:13.355240 run-command.c:1657      run_processes_parallel: done
16:54:13.356215 run-command.c:664       trace: run_command: git maintenance run --auto --no-quiet
16:54:13.432379 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
16:54:13.441166 git.c:444               trace: built-in: git maintenance run --auto --no-quiet

(8)$ git rev-list --count HEAD
624
```
 - What did you expect to occur after running these commands?

This repository have 1663 commits on the master branch, so I would expect after command `(8)`, it would give me a history of 1200 count. In addition (minor), the command `(6)` should had show the commit history count is 601 instead of 624.

 - What actually happened instead?

What happen as so in the result of command `(8)`. The subsequent command `git fetch --deepen` after the first one never expended the history on local repository folder. I stuck at a magic number 624 commits. 

When I run `git log --oneline --all --graph -630`. I could see the history at commit 624 (grafted), and missing 11 commits that connected to the commit show up after the (grafted) one in the log. The git log show as if they are on difference branches, but the fact it is not. Below is the modified git log output to illustrate the issue:
```
...
(#623)    * 73d7accb 
(#624)    * 0f596533 (grafted)
(missing) * c58c6b94 
(missing) * a9a5f3be 
(missing) * 5c5a859a 
(missing) * 8a31199c 
(missing) * f4d2e051 
(missing) * c2b9d195 
(missing) * a09ae088 
(missing) * d87c5423 
(missing) * 0d01be6d 
(missing) * 740d44ee 
(missing) * 491b129e 
(#625)    * 6d8d0d10 
(#626)    * 665faaea 
(#627)    * ed0cfe10 
...
``` 

NOTE: 
1. If the first `git fetch` command `(5)` was `git fetch --deepen 1663`, it would correctly download the full commit history on the master branch. 
2. If I run `git fetch --deepen 100`, I could have it expend the history as normal until the history count reach 624 commits, then it just stuck there.
3. `git fetch --deepen 100` for example, some time it download more than 100 commits into the local. This may be as design, but I just like to point out.
4. My master history are 98% linear (no merge commits). The missing commits are completely linear.

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

Due to the repository own my company - I cannot share, and it also not accessible from the public internet.

**Comment by dscho** [comment]
> Due to the repository own my company - I cannot share, and it also not accessible from the public internet.

In conjunction with not having an [MCVE](http://stackoverflow.com/help/mcve), this makes it very, very hard for others to help.

Maybe you want to spend some time on trying to make it easier to help, e.g. by first testing whether the same problem persists also if you make a full local clone, and then make a shallow clone _from that clone_, then try to deepen that second clone.

If the problem persists, you could try to continue reducing the problem surface e.g. by exporting the repository via `git fast-export --anonymize`, then importing it into a freshly-initialized bare repository, then trying again with a shallow clone of that repository.

If the problem persists, you could then try with a smaller anonymized export (e.g. by fast-exporting only `HEAD~700..HEAD~600`).

Step by step, try to reduce the space in which the bug can hide. If the problem does _not_ persist, it usually means that you removed something very crucial, something that will lead the way to the actual root cause.

**Comment by nhatkhai** [comment]
Yes - it happen the same as local clone too
```
(1a) $ git clone --depth=1 ssh://git@XYZ:7999/ABC/REPO.git
Cloning into 'cricket'...
remote: Enumerating objects: 1787, done.
remote: Counting objects: 100% (1787/1787), done.
remote: Compressing objects: 100% (1446/1446), done.
remote: Total 1787 (delta 358), reused 1498 (delta 278)
Receiving objects: 100% (1787/1787), 6.04 MiB | 3.41 MiB/s, done.
Resolving deltas: 100% (358/358), done.
Updating files: 100% (1685/1685), done.
Filtering content: 100% (701/701), 1015.25 MiB | 1.53 MiB/s, done.

(2a) $ git fetch --deepen 600
fatal: not a git repository (or any of the parent directories): .git

(3a) $ cd REPO

(4a) $ git fetch --deepen 600
remote: Enumerating objects: 5653, done.
remote: Counting objects: 100% (5653/5653), done.
remote: Compressing objects: 100% (1701/1701), done.
remote: Total 5238 (delta 3500), reused 4990 (delta 3285)
Receiving objects: 100% (5238/5238), 9.29 MiB | 3.84 MiB/s, done.
Resolving deltas: 100% (3500/3500), completed with 243 local objects.
remote: Total 0 (delta 0), reused 0 (delta 0)
From ssh://XYZ:7999/ABC/REPO
 * [new tag]         V0.1.2.3 -> V0.1.2.3
 * [new tag]         efg        -> efg
 
(5a) $ git rev-list --count HEAD
624

(6a) $ git fetch --deepen 600
20:18:50.653137 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/bin
20:18:50.662919 git.c:444               trace: built-in: git fetch --deepen 600
20:18:50.669735 run-command.c:664       trace: run_command: unset GIT_PREFIX; GIT_PROTOCOL=version=2 ssh -o SendEnv=GIT_PROTOCOL -p 7999 git@XYZ 'git-upload-pack '\''/ABC/REPO.git'\'''
remote: Total 0 (delta 0), reused 0 (delta 0)
20:18:52.627230 run-command.c:664       trace: run_command: git --shallow-file c:/projects/REPO/.git/shallow.lock unpack-objects --pack_header=2,0
20:18:52.736590 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
20:18:52.745379 git.c:444               trace: built-in: git unpack-objects --pack_header=2,0
20:18:52.758073 run-command.c:664       trace: run_command: git --shallow-file c:/projects/REPO/.git/shallow.lock rev-list --objects --stdin --quiet --alternate-refs
20:18:52.832285 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
20:18:52.841074 git.c:444               trace: built-in: git rev-list --objects --stdin --quiet --alternate-refs
20:18:53.151888 run-command.c:1625      run_processes_parallel: preparing to run up to 1 tasks
20:18:53.151888 run-command.c:1657      run_processes_parallel: done
20:18:53.152864 run-command.c:664       trace: run_command: git maintenance run --auto --no-quiet
20:18:53.242314 exec-cmd.c:237          trace: resolved executable dir: C:/Program Files/Git/mingw64/libexec/git-core
20:18:53.251133 git.c:444               trace: built-in: git maintenance run --auto --no-quiet

(7a) $ git rev-list --count HEAD
624
```

**Comment by dscho** [comment]
> `(1a) $ git clone --depth=1 ssh://git@XYZ:7999/ABC/REPO.git`

Just a few suggestions how to make this "more minimal" of a reproducer:
- Can you reproduce it when cloning with a `file:///<path>` URL, too?
- And what about the restricted export (`HEAD~700..HEAD~600`)?
- What about exporting only the main branch? Does it reproduce with that, still?
- Have you looked at those tags? Is it possible that `--deepen 600` does the correct thing and ensure that `<tag>~600` is fetched and made shallow?

**Comment by nhatkhai** [comment]
@dscho, I will do - I would took me some time to try each of the suggestions. There are 2 suggestions I had not yet done.

**Comment by nhatkhai** [comment]
@dscho - Here we got as promise:
Here is what I do to create anon repos:
```
$ git clone git://git@XYZ/ABC/REPO.git
$ cd REPO
$ git fast-export --anonymize --all > afile
$ cd ..
$ git init anon
$ cd anon
$ git fast-import < ..\REPOS\afile
```

And now I can reproduce the problem
```
$ git clone --depth=1 file://c:/projects/anon ashallow
$ cd ashallow
$ git fetch --deepen 600
$ git fetch --deepen 600
$ git rev-list --count HEAD
624
```

When I try to fast-export for HEAD~700..HEAD~600. I got empty history
```
$ cd REPO
$ git fast-export --anonymize HEAD~700..HEAD~600 > afile2
$ cd ..
$ git init anon2
$ cd anon2
$ git fast-import < ..\REPOS\afile2
$ git log --all
<EMPTY>
```

If you like, I thing I'm confident to share this anonymize repos. I could zip it up and attach it here (1.13MB). Should I ?

**Comment by nhatkhai** [comment]
Now look like I found a break in behavior
```
$ cd anon
$ git fast-export HEAD > afile4
$ cd ..
$ git init anon4
$ cd anon4
$ git fast-import < ..\anon\afile4
$ cd ..
$ git clone --depth 1 file://c:/projects/anon4  -b ref11 shallow4
$ cd shallow4
$ git fetch --deepen 600
$ git rev-list --count HEAD
624

$ git fetch --deepen 600
$ git rev-list --count HEAD
1224

```

**Comment by nhatkhai** [comment]
`fetch --deepen 600 origin HEAD` also got the similar stuck issue when I try this on shallow clone `fille://` with the original full REPOS repository.

**Comment by dscho** [comment]
The next step is probably to generate a shorter commit history. To do that, create a new branch from a previous commit (e.g. `git branch head-400 HEAD~400`) and then export just that ref (i.e. `git fast-export head-400`). Obviously you would want to reduce the depth by the appropriate amount.

If my theory is correct, at some number the `--deepen` call will result in a commit history length that you would expect. This will be where the root cause of this entire issue lies (and I expect it to be a merge commit, potentially a single one that merged in a topic branch of exactly 23 commits).

**Comment by nhatkhai** [comment]
I will try that - but I not sure it give use what we looking for. Because as I show you when I do the `git fast-export HEAD`, it work fine as show in https://github.com/git-for-windows/git/issues/3192#issuecomment-824533095

**Comment by nhatkhai** [comment]
As I predicted with the head-400 try - It work fine. `git fetch --deepen 200` keep continue expand the history exactly 200 each execution.

The `git fast-export <branch>` must had remove the interaction with other tags/branches. Look like the interaction with other tags causing the issues.

**Comment by nhatkhai** [comment]
Also the full history git log look like this:
```
$ git log --all --graph --merges
* f6d52501 subject 1338

```

**Comment by nhatkhai** [comment]
Interesting that -if it redo the fetch from clone depth=1. Here what I got:
```
$ git fetch --deepen 334 && git rev-list --count HEAD
335

$ git fetch --deepen 1 && git rev-list --count HEAD
336

$ git fetch --deepen 1 && git rev-list --count HEAD
338

$ git fetch --deepen 1 && git rev-list --count HEAD
340

$ git fetch --deepen 1 && git rev-list --count HEAD
342

```
After the first every merge, it deepen by two for deepen=1. Even they are on the same branch. No tag, no branch point to this section of commits.

**Comment by nhatkhai** [comment]
`fetch --deepen 600` stuck at 624 deep in history after the first call.
`fecth --deepen 597` stuck at 621 deep in history after the first call.

**Comment by nhatkhai** [comment]
Back to `head-400`. I also try this:
```
$ git clone --depth 1 file://c:/projects/removeme/anon shallow -b head-400
$ git fetch --deepen 200 && git rev-list --count HEAD
201

$ git fetch --deepen 200 && git rev-list --count HEAD
401

$ git fetch --deepen 200 && git rev-list --count HEAD
601

$ git fetch --deepen 200 && git rev-list --count HEAD
801

$ git fetch --deepen 200 && git rev-list --count HEAD
1001

$ git fetch --deepen 200 && git rev-list --count HEAD
1201

$ git fetch --deepen 600 && git rev-list --count HEAD
1251  (reached the end, no stuck)

$ git branch -a
* head-400
  remotes/origin/head-400
```

**Comment by nhatkhai** [comment]
I also try
```
$ git clone --depth 1 file://c:/projects/removeme/anon shallow -b head-400
$ cd shallow

$ git fetch --deepen 200 && git rev-list --count HEAD
From file://c:/projects/anon
 * [new tag]         ref6       -> ref6
201

$ git fetch --deepen 200 && git rev-list --count HEAD
201
```

**Comment by dscho** [comment]
I am getting the feeling that I should clarify one thing. `git fetch --deepen 600` does _not_ fetch 600 more commits. Instead, it makes sure that the fetched refs' commit graphs are extended so that each strand is 600 commits long. For example, if you have a commit history like this:

```
* merge
* \ left
| * right
|/
* base
```

and you fetched `merge` before, and it was shallow, and you deepen by 1, then _both_ `left` and `right` are fetched.

Maybe a better example is a script that you can play with:

```sh
#!/bin/sh

rm_repo () {
	test -d "$1" || return 0

	git -C "$1" fsmonitor--daemon --stop
	rm -rf "$1"
}

rm_repo upstream &&
git init upstream &&
git -C upstream commit --allow-empty -m root &&
git -C upstream commit --allow-empty -m right &&
git -C upstream commit --allow-empty -m second-right &&
git -C upstream reset --hard HEAD^{/root} &&
git -C upstream commit --allow-empty -m left &&
git -C upstream merge -m merge HEAD@{2} &&

rm_repo clone &&
git clone --depth 1 file://$PWD/upstream clone &&
git -C clone log --oneline --graph &&
git -C clone fetch --deepen 1 &&
git -C clone log --oneline --graph &&
git -C clone fetch --deepen 1 &&
git -C clone log --oneline --graph
```

**Comment by nhatkhai** [comment]
Do I give you enough info for debugging? Let me known if you need anything else.

**Comment by siprbaum** [comment]
@nhatkhai My gut feeling when reading this thread is that there is a missunderstanding of the deepen feature on your end.
Try re-reading the example dscho gave above. Perhaps this clarifies it?

**Comment by siprbaum** [comment]
On second thought, you mentioned that you have grafted the history. Did you do this with `git replace` or by adding the graft
in `.git/info/grafts`? It wouldn't surprise me if the deepening logic gets confused in the present of grafts or replace objects.

**Comment by nhatkhai** [comment]
@siprbaum , @dscho - Please note my problem statement have nothing todo with what I expected how many commits deepen work. It is fact "stuck" is the problem.

**Comment by nhatkhai** [comment]
> On second thought

No git replace, and no manual touch stuff in .git folder. I had show exact all command I use started from clone in all of the above comment.

**Comment by dscho** [comment]
> It is fact "stuck" is the problem.

Maybe I don't really understand your problem, then. When I read that you got stuck on a fetch, I actually assumed that the fetch never finished, so I actually _know_ that we use the word "stuck" very differently.

But since you successfully anonymized the repository, maybe you can push that anonymized version (using `--no-data` to be sure) somewhere. I am still not having any concrete idea what the problem is, apparently, because you seem not to be concerned with the fact that the number of commits is different from 600 (which I had assumed you meant by "it got stuck").

**Comment by nhatkhai** [comment]
I have the same meaning of "stuck" - the fetch never able to fetch all the commits - as shown.
I try to try so many cases for you as use asked to show that they are still stuck - except when you do "git fast-export HEAD", or "git clone --depth 1 .... -b HEAD-400". Other than that all the normal use cases did got stuck.

**Comment by nhatkhai** [comment]
I had attached the `anon` zipped folder -- A result of full fast-export ran from https://github.com/git-for-windows/git/issues/3192#issuecomment-824497112:
[anon.tar.gz](https://github.com/git-for-windows/git/files/6379135/anon.tar.gz)

May be using this, and check it out yourself  would clear thing up.


I also go back and add a bold text to clarify when I try to show you it was "stuck".

**Comment by siprbaum** [comment]
So I tried to to reproduce. This is what I did:
```
cd /c/temp
tar -xvzf anon.tar.gz
git clone --depth=1 file://c:/temp/anon ashallow
cd ashallow
git fetch --deepen 600
```
This works, nothing stuck here. 

So please provide a [Minimal and Verifiable Reproducer](http://stackoverflow.com/help/mcve)
based on the anon.tar.gz you have given, e.g. in a receipe like I have done.
I know that there are quite a few other posts by you in this thread, but it is really hard to follow which
of the operations you gave are just your train of thought or the reproducer.

**Comment by nhatkhai** [comment]
@siprbaum, had you run a 2nd fetch yet?

**Comment by siprbaum** [comment]
D'oh. You are correct. Any subsequent `git fetch --deepen N`  where `N` is an integer (I tried with N=1 and 600) results in
```
remote: Total 0 (delta 0), reused 0 (delta 0), pack-reused 0
```
So that is what you mean by **stuck**, as no more items are fetched.

**Comment by nhatkhai** [comment]
Yep.

**Comment by dscho** [comment]
Closing as stale.
